src/stream/src/executor/dynamic_filter.rs

```diff
@@ -41,7 +44,210 @@ use super::{
 };
 use crate::common::table::state_table::StateTable;
 use crate::common::StreamChunkBuilder;
-use crate::executor::expect_first_barrier_from_aligned_stream;
+use crate::executor::{expect_first_barrier_from_aligned_stream, StreamExecutorResult, Watermark};
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+enum DynamicFilterWatermarkCacheEntry {
+    Empty,
+    Evicted,
+    // We need to store pk, in case we encounter `Delete`s.
+    // Then we can remove the Entry.
+    Smallest { pk: OwnedRow, value: ScalarImpl },
+}
+
+/// Watermark Filter Cache
+/// Used to avoid issuing delete range requests for state clean-up by watermark.
+/// It tracks the lowest value in the LHS table.
+///
+/// On receiving new LHS watermark:
+/// Nothing.
+///
+/// On receiving new RHS watermark:
+/// 1. If empty, unused clean hint -> None.
+/// 1. If watermark < lowest value
+///    We don't need to issue delete range,
+///    because are no values lower than lowest value.
+///    Therefore, no values before watermark, and no values to clean.
+///    We can set `unused_clean_hint` to None.
+/// 2. Else watermark => lowest value
+///    Need to issue delete range.
+///    We can just issue delete range for (`lowest_value`, watermark].
+///    Set `unused_clean_hint` to `watermark`.
+///    TODO(kwannoel): Optimization: If we move this to state table, we can set the delete range
+/// lower bound.
```

Comment:
Need to update this section. Does not seem really accurate. It should be last cleaned watermark, rather than new watermark.
New watermark is not cleaned yet.